src/loader/x86_64/elf/mod.rs

```diff
@@ -0,0 +1,468 @@
+            struct_util::read_struct(kernel_image, &mut ehdr).map_err(|_| Error::ReadElfHeader)?;
+        }
+
+        Self::validate_header(&ehdr)?;
+
+        if let Some(addr) = highmem_start_address {
+            if (ehdr.e_entry as u64) < addr.raw_value() {
+                return Err(Error::InvalidEntryAddress.into());
+            }
+        }
+
+        let mut loader_result: KernelLoaderResult = Default::default();
+
+        // Address where the kernel will be loaded.
+        loader_result.kernel_load = match kernel_start {
+            Some(start) => GuestAddress(
+                start
+                    .raw_value()
+                    .checked_add(ehdr.e_entry as u64)
+                    .ok_or(KernelLoaderError::InvalidKernelStartAddress)?,
+            ),
+            None => GuestAddress(ehdr.e_entry as u64),
+        };
+
+        kernel_image
+            .seek(SeekFrom::Start(ehdr.e_phoff))
+            .map_err(|_| Error::SeekProgramHeader)?;
+        let phdrs: Vec<elf::Elf64_Phdr> = unsafe {
+            // Reading the structs is safe for a slice of POD structs.
+            struct_util::read_struct_slice(kernel_image, ehdr.e_phnum as usize)
+                .map_err(|_| Error::ReadProgramHeader)?
+        };
+
+        // Read in each section pointed to by the program headers.
+        for phdr in &phdrs {
+            if phdr.p_type != elf::PT_LOAD || phdr.p_filesz == 0 {
+                if phdr.p_type == elf::PT_NOTE {
+                    // This segment describes a Note, check if PVH entry point is encoded.
+                    loader_result.pvh_entry_addr = parse_elf_note(phdr, kernel_image)?;
+                }
+                continue;
+            }
+
+            kernel_image
+                .seek(SeekFrom::Start(phdr.p_offset))
+                .map_err(|_| Error::SeekKernelStart)?;
+
+            // if the vmm does not specify where the kernel should be loaded, just
+            // load it to the physical address p_paddr for each segment.
+            let mem_offset = match kernel_start {
+                Some(start) => start
+                    .checked_add(phdr.p_paddr as u64)
+                    .ok_or(Error::InvalidProgramHeaderAddress)?,
+                None => GuestAddress(phdr.p_paddr as u64),
+            };
+
+            guest_mem
+                .read_exact_from(mem_offset, kernel_image, phdr.p_filesz as usize)
+                .map_err(|_| Error::ReadKernelImage)?;
+
+            loader_result.kernel_end = mem_offset
+                .raw_value()
+                .checked_add(phdr.p_memsz as GuestUsize)
+                .ok_or(KernelLoaderError::MemoryOverflow)?;
+        }
+
+        // elf image has no setup_header which is defined for bzImage
+        loader_result.setup_header = None;
+
+        Ok(loader_result)
+    }
+}
+
+fn parse_elf_note<F>(phdr: &elf::Elf64_Phdr, kernel_image: &mut F) -> Result<Option<GuestAddress>>
```

Comment:
This function was not documented before. Would it make sense to add some documentation?

Reply:
Thanks for catching these 2, I'll take the opportunity to document them. They're not public facing so I'll skip doctests but I agree that they could use some documentation :+1:

Reply:
This one is on me :). The description for `parse_elf_note()` should be something like:

"Examines a supplied ELF program header of type PT_NOTE to determine if it contains an entry of type XEN_ELFNOTE_PHYS32_ENTRY (0x12). Notes of this type encode a physical 32-bit entry point address into the kernel which is used when launching guests in 32-bit (protected) mode with paging disabled, as described by the PVH boot protocol. 

On a successful execution, parse_elf_note() returns an optional type containing the encoded entry point address, or None if no entries of the XEN_ELFNOTE_PHYS32_ENTRY type are found in the note header provided as the first argument."

Or something to that effect..

Reply:
Thanks @aljimenezb, I added the documentation.